doc/specs/#1595 - Suggestions UI/Suggestions-UI.md

```diff
@@ -0,0 +1,734 @@
+
+### Implementation Details
+
+#### Fork the Command Palette
+
+We're largely going to start with the Command Palette to build the Suggestions
+UI[[1](#footnote-1)]. The Command Palette is already a control we've built for displaying a
+transient list of commands and dispatching them to the rest of the app.
+
+Currently, the Command Palette is a single static control, at the top-center of
+the Terminal window, and occupying a decent portion of the screen. For the
+Suggestions UI, we'll instead want to make sure that the control appears
+relative to the current cursor position.
+
+We'll start by taking the command palette, and copying it over to a new control.
+This will allow us to remove large chunks of code dealing with different modes
+(i.e. the tab switcher), and code dealing with prefix characters to switch
+modes.
+
+We'll need to make some small modifications to enable the Suggestions UI to
+* work as a text cursor-relative control
+* exist as a Flyout outside the bounds of the Terminal window
+* If the Suggestions UI is too close to the bottom of the screen, we'll need it to open
+  "upwards", with the search box at the _bottom_ and the list extending above it
+* prevent it from switching to command-line mode
+* display tooltips / `TeachingTip`s / some secondary flyout with a description
+  of the suggestion (if provided)
+
+#### Completion sources
+
+The Suggestions UI will support suggestions from a variety of different
+"sources". As an example, consider the following actions:
+
+```json
+        { "command": { "action":"suggestions", "source": "commandHistory" } },
+        { "command": { "action":"suggestions", "source": "directoryHistory" } },
+
+        { "command": { "action":"suggestions", "source": "tasks" } },
+        { "command": { "action":"suggestions", "source": "local" } },
+
+        { "command": { "action":"suggestions", "source": ["local", "tasks", "commandHistory"] } },
+
+        { "command": { "action":"suggestions", "source": "Microsoft.Terminal.Extensions.BufferComplete" } },
+```
+
+Each of these `suggestions` actions would open the Suggestions UI with a
+different set of actions.
+
+* `commandHistory`: Use commands from this session, as identified via shell
+  integration. This won't be able to return any suggestions if the user has not
+  configured their shell to support shell integration sequences yet.
+* `directoryHistory`: Populate the list with a series of `cd {path}` commands,
+  where the paths are populated via shell integration. Paths are in MRU order.
+* `tasks`: Populate the list with all `sendInput` actions in the user's settings
+  file. The command structure should remain unchanged. For example, if they have
+  `sendInput` actions nested under a "git" command, then the "git" entry will
+  remain in this tasks view with their `sendInput` actions nested inside it. For
+  more details, see the [Tasks] spec.
+* `local`: Populate the list with tasks that are located in the CWD, in a file

```

Comment:
nit: call out association with shell integration; that's the only way we'll know the CWD

How do we handle "project" style directory trees? For example:

```
PS> cd c:\src\terminal
# Yay, it loaded c:\src\terminal\.wt.json
PS> cd tools
# oh no, tools doesn't contain a .wt.json
```

I bet this is part of the tasks spec. I should read that next. Hold fire on this comment.

Reply:
Yea that's in the tasks spec, and kinda only vaguely referenced here for completion's sake.